### ISSUE TITLE
Proposal: Have Logstash use rollover indices by default

### ISSUE THREAD

**Issue by untergeek** [logstash-plugins/logstash-output-elasticsearch#636]
If we can also make Logstash handle rollover, we can use rollover indices by default.

The basic proposal would be to:

1. This plugin uses rollover by default
2. To keep things roughly the same, we can default to daily rollover, even preserving the `YYYY.MM.dd` naming scheme, but this doesn't necessarily have to be so.
3. Allow users to specify doc count and/or age conditions for Rollover
4. Add a background thread that runs the Rollover API call at `rufus` scheduler intervals (which could also be a configurable parameter)

Thoughts?  Different ideas?

**Comment by ppf2** [comment]
I like the idea of doing this natively in Logstash so it doesn't have to outsource to the curator just to roll the indices.   I think Step 2 will be nice - to preserve the naming scheme of the underlying indices if possible so that existing curator scripts that match indices by index name patterns will continue to work?  Otherwise, users will have to change their scripts to filter by source>creation_date when rollover is implemented? (https://www.elastic.co/guide/en/elasticsearch/client/curator/5.1/fe_source.html#_literal_creation_date_literal_based_ages_2)

**Comment by suyograo** [comment]
> Allow users to specify doc count and/or age conditions for Rollover

Big +1 on this.

**Comment by jordansissel** [comment]
I like this proposal. I'm thinking we can't really achieve "preserve the naming scheme" because of a technicality -- clocks.

This proposal changes the clock source that determines the index an event will be delivered to.

* Current behavior: The Event's timestamp (presumably based on when the event happened)
* Proposed behavior: Elasticsearch's system clock (the time on the Elasticsearch instance)

The net impact of this is that users will no longer be able to rely on the index name for limit searches by time range. For example, you can have a situation where 7-day old logs are submitted resulting in:

* Current behavior: The 7-day old logs are placed in the index named for 7 days ago
* Proposed behavior: The 7-day old logs are placed in the "today" index

This can be really confusing if a user queries the `logstash-2017.08.10` index but because of this situation some of the logs for this day ("2017-08-10") are in the `logstash-2017.08.17` index.

The behavior I describe above assumes folks are using the `date` filter or are otherwise setting the Logstash `@timestamp` field correctly. 

Thoughts on how to work around this? This kind of problem is infrequent and rare, but can be incredibly frustrating for someone who encounters it.

**Comment by dustin-decker** [comment]
At minimum I'd like Logstash to be able to trigger the rollover API because it is responsible for indexing, and if an external tool fails to call the rollover then the index will blow up.

**Comment by jordansissel** [comment]
@dustin-decker This is true already with Curator, so the problem doesn't get worse, imo, regardless of who makes the rollover call. By this, I mean, many folks use the excellent Curator ("external tool") to do index lifecycle management -- deleting older indices, etc. And with your scenario, this external tool could fail to run, for whatever reason, and you won't end up deleting older indices, and you end up running out of disk space the same way you might by not failing to call rollover api.

So to me, having Logstash call it or something else call it is more a convenience/operational concern, not about points of failure, because the points of failure already exist today without rollover.
